sdk/storage/src/cloud_location.rs

```diff
@@ -44,28 +88,123 @@ impl CloudLocation {
+                    return match name.as_str() {
+                        // Seems like "AzurePublicCloud" is used in some environments
+                        AZURE_CLOUD | AZURE_PUBLIC_CLOUD => CloudLocation::Public {
+                            account: account.clone(),
+                            credentials: credentials.clone(),
+                        }
+                        .url(service_type),
+                        AZURE_US_GOV => CloudLocation::USGov {
+                            account: account.clone(),
+                            credentials: credentials.clone(),
+                        }
+                        .url(service_type),
+                        AZURE_CHINA_CLOUD => CloudLocation::China {
+                            account: account.clone(),
+                            credentials: credentials.clone(),
+                        }
+                        .url(service_type),
+                        _ => {
+                            return Err(azure_core::Error::with_message(
+                                azure_core::error::ErrorKind::Other,
+                                || {
+                                    format!(
+                                        "Auto-detect encountered an invalid cloud name, allowed values are: {AZURE_CLOUD}, {AZURE_PUBLIC_CLOUD}, {AZURE_US_GOV}, {AZURE_CHINA_CLOUD}.",
+                                    )
+                                },
+                            ));
+                        }
+                    };
+                } else {
+                    return Err(azure_core::Error::with_message(
+                        azure_core::error::ErrorKind::Other,
+                        || {
+                            "Auto-detect could not find a cloud name from the current environment."
+                                .to_string()
+                        },
+                    ));
+                }
+            }
         };
         Ok(url::Url::parse(&url)?)
     }
 
+    /// Returns the storage credentials for this cloud location,
+    ///
     pub fn credentials(&self) -> &StorageCredentials {
         match self {
-            CloudLocation::Public { credentials, .. } => credentials,
-            CloudLocation::China { credentials, .. } => credentials,
+            CloudLocation::Public { credentials, .. }
+            | CloudLocation::China { credentials, .. }
+            | CloudLocation::USGov { credentials, .. }
+            | CloudLocation::Custom { credentials, .. }
+            | CloudLocation::AutoDetect { credentials, .. } => credentials,
             CloudLocation::Emulator { .. } => &EMULATOR_CREDENTIALS,
-            CloudLocation::Custom { credentials, .. } => credentials,
+        }
+    }
+
+    /// Finds the cloud name, first by environment variable, then by parsing the current user's $HOME/.azure/config file
+    ///
+    fn find_cloud_name() -> Option<String> {
+        if let Ok(name) = std::env::var("AZURE_CLOUD_NAME") {
```

Comment:
Is `AZURE_CLOUD_NAME` used by any existing Azure SDKs? https://github.com/Azure/azure-sdk-for-net/search?q=AZURE_CLOUD_NAME comes up empty.

Reply:
I couldn't find anything either, but apparently, it's used by other libraries for cloud detection. I actually think this env variable might make more sense, https://github.com/Azure/azure-sdk-for-net/blob/eca8faf3c868bbb75ed8a104661bc77f5a976c9a/sdk/identity/Azure.Identity/src/EnvironmentVariables.cs#L30